cpp/src/fil/infer.cu

```diff
@@ -531,45 +555,67 @@ template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
 __global__ void infer_k(storage_type forest, predict_params params) {
   extern __shared__ char smem[];
   float* sdata = (float*)smem;
+  int sdata_stride = params.sdata_stride();
+  int rows_per_block = NITEMS << params.log2_threads_per_tree;
   int num_cols = params.num_cols;
-  for (size_t block_row0 = blockIdx.x * NITEMS; block_row0 < params.num_rows;
-       block_row0 += NITEMS * gridDim.x) {
-    size_t num_input_rows = min((size_t)NITEMS, params.num_rows - block_row0);
+  int thread_row0 = NITEMS * modpow2(threadIdx.x, params.log2_threads_per_tree);
+  for (long long block_row0 = blockIdx.x * rows_per_block;
+       block_row0 < params.num_rows; block_row0 += rows_per_block * gridDim.x) {
+    int block_num_rows =
+      max(0, (int)min((long long)rows_per_block, params.num_rows - block_row0));
     const float* block_input = params.data + block_row0 * num_cols;
     if (cols_in_shmem) {
       // cache the row for all threads to reuse
-      size_t feature = 0;
+      // 2021: latest SMs still do not have >262KB of shared memory/block required to
+      // exceed the unsigned short
 #pragma unroll
-      for (feature = threadIdx.x; feature < num_input_rows * num_cols;
-           feature += blockDim.x)
-        sdata[feature] = block_input[feature];
+      for (unsigned short input_idx = threadIdx.x;
+           input_idx < block_num_rows * num_cols; input_idx += blockDim.x) {
+        // assuming here that sdata_stride == num_cols + 1
+        // then, idx / num_cols * sdata_stride + idx % num_cols == idx + idx / num_cols
+        unsigned short sdata_idx =
+          sdata_stride == num_cols
+            ? input_idx
+            : input_idx + input_idx / (unsigned short)num_cols;
```

Comment:
Is this required because of the shared memory padding? Is the shared memory padding required to avoid bank conflicts? Please add comments explaining this.